src/partition/repart.c

```diff
@@ -3423,6 +3500,9 @@ static int partition_populate_filesystem(Partition *p, const char *node) {
                 if (mount_nofollow_verbose(LOG_ERR, node, fs, p->format, MS_NOATIME|MS_NODEV|MS_NOEXEC|MS_NOSUID, NULL) < 0)
                         _exit(EXIT_FAILURE);
 
+                if (do_make_subvolumes(p, fs) < 0)
+                        _exit(EXIT_FAILURE);
+
                 if (do_copy_files(p, fs) < 0)
                         _exit(EXIT_FAILURE);
 
```

Comment:
this interesting. first making the subvols, then copying data in, and then making the dirs.

we make the dirs last because of perms: we'd rather use the right perms from the copy tree than the "vanilla" perms we ourselves create here.

now, subvolumes we *have* to create before the copying in. but i wonder what that means for the perms. I think our copy routine currently is not able to copy just perms/xattrs/… from the source to the target if the target already exists.

So maybe we should add a new flag to the copy.c logic: COPY_UPDATE_DIR_ATTRS or so? if that flag is set, we'll copy over all permissions/ownership/acls/xattrs/chattrs and so on from the source to the target, in case the target already exists.

we'd then set this flag unconditionally here.

and i'd really only do this for dirs, not for the other inode types.

and then, to keep this systematic i#d probably move the do_make_directories() after the do_make_subvolumes() to make this less surprising and since it then doesn't matter anymore.

Reply:
well, if we have multiple trees to copy in it's not always clear whether the first or the last tree in the list should win with its permissions. You argue for "first wins", I propose "last wins", because I see  MakeSubvolume= as just another tree, and I want that overriden...

but maybe the solution is to introduce the COPY_UPDATE_DIR_ATTRS flag as I proposed and then set it for the first tree to copy in, but not for the others? that sound like a reasonable approach?

Reply:
That seems overly restrictive, let's say we have a root filesystem that is composed from different directories on the host filesystem that aren't yet set up in the expected order, e.g. we have /a/b/c/usr which should go to /usr and /b/d/f/etc which should go to /etc. If we only set the flag for the first tree we copy in, we'll only override attributes for /usr but not for /etc. Perhaps we should make this user configurable instead?